traitlets/traitlets.py

```diff
@@ -644,6 +648,17 @@ def hold(*a):
                 for v in cache.values():
                     self._notify_trait(*v)
 
+    @contextlib.contextmanager
+    def _allow_write(self):
```

Comment:
Having state on the trait for this seems like the wrong way to do this. E.g. any on_change callbacks for the traitlet will be able to modify it, because they'll be called while it's 'unlocked'. Likewise, if there's more than one thread that might try to change it, some changes that should raise an error could randomly succeed.

I'm not convinced this belongs in traitlets at all, but if it does, I think the design should be fundamentally different. For instance, a main interface using properties that is always read only, and a different 'back door' interface that can be used to update the stored value.